examples/mixture_models/dirichlet_mixture_of_multinomials.ipynb

```diff
@@ -0,0 +1,1815 @@
+{
```

Comment:
<p>Simplify to "Thankfully, we only get one warning related to inefficient sampling"?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc-examples/pull/18/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Reply:
<p>Again, it's because Metropolis isn't as good as NUTS to spot and warn about problematic geometries in the posterior. I'd reformulate as: "we only get one warning now, so it <em>seems</em> that our sampling was more efficient".</p><p>Also, I think you should show how this model runs with NUTS, so that people can see what inefficient sampling looks like</p>

Reply:
<blockquote>Also, I think you should show how this model runs with NUTS, so that people can see what inefficient sampling looks like</blockquote><p> </p><p>Unfortunately, NUTS fails entirely with:</p><p> </p><pre class="ql-syntax" spellcheck="false">---------------------------------------------------------------------------
RemoteTraceback                           Traceback (most recent call last)
RemoteTraceback: 
"""
Traceback (most recent call last):
  File "~/Projects/pymc3/pymc3/parallel_sampling.py", line 191, in _start_loop
    point, stats = self._compute_point()
  File "~/Projects/pymc3/pymc3/parallel_sampling.py", line 216, in _compute_point
    point, stats = self._step_method.step(self._point)
  File "~/Projects/pymc3/pymc3/step_methods/arraystep.py", line 276, in step
    apoint, stats = self.astep(array)
  File "~/Projects/pymc3/pymc3/step_methods/hmc/base_hmc.py", line 159, in astep
    raise SamplingError("Bad initial energy")
pymc3.exceptions.SamplingError: Bad initial energy
"""

The above exception was the direct cause of the following exception:

SamplingError                             Traceback (most recent call last)
SamplingError: Bad initial energy

The above exception was the direct cause of the following exception:

ParallelSamplingError                     Traceback (most recent call last)
&lt;ipython-input-22-16dea9c651bc&gt; in &lt;module&gt;
      1 with model_multinomial:
----&gt; 2     trace_multinomial = pm.sample(
      3         draws=int(5e3), chains=4, return_inferencedata=True
      4     )

~/Projects/pymc3/pymc3/sampling.py in sample(draws, step, init, n_init, start, trace, chain_idx, chains, cores, tune, progressbar, model, random_seed, discard_tuned_samples, compute_convergence_checks, callback, jitter_max_retries, return_inferencedata, idata_kwargs, mp_ctx, pickle_backend, **kwargs)
    555         _print_step_hierarchy(step)
    556         try:
--&gt; 557             trace = _mp_sample(**sample_args, **parallel_args)
    558         except pickle.PickleError:
    559             _log.warning("Could not pickle model, sampling singlethreaded.")

~/Projects/pymc3/pymc3/sampling.py in _mp_sample(draws, tune, step, chains, cores, chain, random_seed, start, progressbar, trace, model, callback, discard_tuned_samples, mp_ctx, pickle_backend, **kwargs)
   1473         try:
   1474             with sampler:
-&gt; 1475                 for draw in sampler:
   1476                     trace = traces[draw.chain - chain]
   1477                     if trace.supports_sampler_stats and draw.stats is not None:

~/Projects/pymc3/pymc3/parallel_sampling.py in __iter__(self)
    477 
    478         while self._active:
--&gt; 479             draw = ProcessAdapter.recv_draw(self._active)
    480             proc, is_last, draw, tuning, stats, warns = draw
    481             self._total_draws += 1

~/Projects/pymc3/pymc3/parallel_sampling.py in recv_draw(processes, timeout)
    357             else:
    358                 error = RuntimeError("Chain %s failed." % proc.chain)
--&gt; 359             raise error from old_error
    360         elif msg[0] == "writing_done":
    361             proc._readable = True

ParallelSamplingError: Bad initial energy
</pre><p> </p><p>I messed around with <code>target_accept</code>  and <code>testval</code>,  but I couldn't get all 4-chains to finish without numerical issues.</p>

Reply:
<p>I haven't drilled down into the NUTS sampling errors, but I'm guessing that we're getting logp's or dlogp's of <code>-inf</code> due to the large deviation from the multinomial predictive density (e.g. maybe <code>observed_counts[0, 4] == 4</code> is just too unbelievable?).</p><p> </p><p>Since Metropolis either (A) rejects jumps into regions of the parameter space where logp is <code>-inf</code>, or (B) doesn't know about the numerically unstable gradient, it doesn't fail the same way NUTS does.</p><p> </p>